# Null_models_I_and_II

 Simulating annealing based strategies to randomize sequences data.
 
 A complete description of the algorithm can be found at (Edwin Rodriguez Horta, Weigt M.,Phylogenetic correlations have limited effect on coevolution-based contact prediction in proteins).

# Install
To install the package under version >= 1.0 use the package manager 

(v1.?) pkg> add https://github.com/ed-rodh/Null_models_I_and_II

# Overview
There is two strategies:
 
  + Null model I: ramdomizes the input MSA by conserving the single columns statistics.
 
  + Null model II: randomizes the input MSA preserving both single columns statistics and pairwise hamming distances between sequences.
   
In this package each entry of the MSA is a letter in A={-, A, C, ..., Y} being either one of the 20 amino acids, or an alignment gap, “–”, representing amino acid insertions or deletions. The total alphabet size is q = |A| = 21. The mapping of this alignments to matrix of numbers assume {-, A, C, ..., Y} -> {1, 2, 3,..., 21}. 

# Usage
Load the julia package by

 ```julia
 julia> using Null_models_I_and_II
 ```
This software provide two main functions ```sample_from_Null_model_I(infile::String;outfile="",Temp=10.0)``` and ```sample_from_Null_model_II(infile::String;outfile="",schuffle=1,Temp=10.0)``` which are in charge of generating new sequence data according to null models I and II respectively. Function ```sample_from_Null_model_I()``` is faster. 

 + Both functions take as input a file with protein's multiple sequence alignments (MSA)  in fasta format. 
 + Optional parameter   "outfile"    name text file for the output alignment exported in fasta format.
 + Parameter "Temp" stands for the  temperature in Monte Carlo sampling to shuffling the sequences data (is recommended to set Temp>1).
 + Parameter 'schuffle' allows (schuffle=1) or not (schuffle=0) to initialize the sampling strategy of the null modell II from sequences data generated by null model I (is recommended schuffle=1). 

We also provide other usefull functions:
 + ```translate_fasta_to_num_matrix(msa_fasta::String,...)``` to read the MSA as fasta file and transform to a MSA as a matrix of numbers.
 + ```transform_MSA_fasta(msa::Array{Int64,2},...)``` transform MSA from numbers to letters.
 + ```export_fasta_file(outfile::String,msa::Array{Char,2})``` export a MSA in fasta format from a matrix of letters.
 + ```PairwiseHammingDist(sample::Array{Int64,2})``` compute the sequences pairwise hamming distance from sequences of numbers stored in the matrix "sample".

# Output

The functions output a type Array{Char,2} wich describe a matrix of aminoacids (letters), then if outfile is not empty  the output alignment is exported as a text file in fasta format.

=> For a quick test we recommend the alignment included in Null_models_I_and_II/data_test
# Requirements

The code requires at least Julia version 1.0 or later.

